hcl/eval/partial.go

```diff
@@ -0,0 +1,1388 @@
+	case hclsyntax.TokenOBrack, hclsyntax.TokenDot:
+		err := e.evalAcessors()
+		if err != nil {
+			return err
+		}
+		e.commit()
+	}
+
+	e.multiline--
+
+	return nil
+}
+
+func (e *engine) evalGetAttr() error {
+	e.newnode()
+	if e.peek().Type != hclsyntax.TokenDot {
+		panic("expected . IDENT (getAttr)")
+	}
+
+	e.emit()
+	e.emitnlparens()
+	e.emitComments()
+	tok := e.peek()
+	if tok.Type == hclsyntax.TokenIdent ||
+		tok.Type == hclsyntax.TokenNumberLit {
+		e.emit()
+	} else {
+		panic(sprintf("expect an IDENT or number: %s %t", e.peek().Type, e.multiline))
+	}
+
+	return nil
+}
+
+func (e *engine) evalIdent() error {
+	e.newnode()
+	if !e.canEvaluateIdent() {
+		return errorf("malformed code")
+	}
+
+	tok := e.peek()
+	if tok.Type != hclsyntax.TokenIdent {
+		panic(errorf("evalIdent: unexpected token '%s' (%s)", tok.Bytes, tok.Type))
+	}
+
+	next := e.peekn(e.skip(1))
+	switch next.Type {
+	case hclsyntax.TokenDot:
+		err := e.evalVar()
+		if err != nil {
+			return err
+		}
+		e.commit()
+	case hclsyntax.TokenOParen:
+		err := e.evalFuncall()
+		if err != nil {
+			return err
+		}
+		e.commit()
+	default:
+		panic("ident cannot be evaluated")
+	}
+
+	return nil
+}
+
+func (e *engine) evalList() error {
+	e.newnode()
+	tok := e.peek()
+	if tok.Type != hclsyntax.TokenOBrack {
+		panic("bug")
+	}
+
+	e.multiline++
+
+	e.emit()
+	e.emitnl()
+	for e.hasTokens() && e.peek().Type != hclsyntax.TokenCBrack {
+		err := e.evalExpr()
+		if err != nil {
+			return err
+		}
+
+		e.commit()
+		tok := e.peek()
+		if tok.Type == hclsyntax.TokenComma {
+			e.emit()
+		}
+		e.emitnl()
+	}
+
+	if !e.hasTokens() {
+		panic("malformed list")
+	}
+
+	tok = e.peek()
+	if tok.Type != hclsyntax.TokenCBrack {
+		panic(errorf("malformed list, unexpected %s", tok.Bytes))
+	}
+
+	e.multiline--
+
+	e.emit()
+	return nil
+}
+
+func (e *engine) evalObject() error {
+	e.newnode()
+	tok := e.peek()
+	if tok.Type != hclsyntax.TokenOBrace {
+		panic("bug")
+	}
+
+	e.emit()
+	e.emitnl()
+	for e.hasTokens() && e.peek().Type != hclsyntax.TokenCBrace {
+		err := e.evalExpr()
+		if err != nil {
+			return err
+		}
+
+		e.commit()
+		e.emitnl()
+		tok = e.peek()
+		if tok.Type != hclsyntax.TokenEqual && tok.Type != hclsyntax.TokenColon {
+			panic(errorf("evalObject: unexpected token '%s' (%s)", tok.Bytes, tok.Type))
+		}
+
+		e.emit()
+		err = e.evalExpr()
+		if err != nil {
+			return err
+		}
+
+		e.commit()
+		tok = e.peek()
+		if tok.Type == hclsyntax.TokenComma {
+			e.emit()
+		}
+
+		e.emitnl()
+	}
+
+	if !e.hasTokens() {
+		panic("malformed object")
+	}
+
+	tok = e.peek()
+	if tok.Type != hclsyntax.TokenCBrace {
+		panic(errorf("malformed object, unexpected %s", tok.Bytes))
+	}
+
+	e.emit()
+	return nil
+}
+
+func (e *engine) evalForExpr(matchOpenType, matchCloseType hclsyntax.TokenType) error {
+	_, thisNode := e.newnode()
+	// { | [
+	tok := e.peek()
+	if tok.Type != matchOpenType {
+		panic(sprintf("evalForExpr: malformed `for` expression: %s", tok.Bytes))
```

Comment:
```suggestion
		panic(errorf("evalForExpr: malformed `for` expression: %s", tok.Bytes))
```

Or we could go consistently with sprintf.